### ISSUE TITLE
Figure5

### ISSUE THREAD

**Issue by Curiosity-Mars** [Curiosity-Mars/total-testosterone#7]
import os
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler
from sklearn.cluster import KMeans
from scipy.stats import spearmanr

# -------------------------
# 0) Colab download helper
# -------------------------
try:
    from google.colab import files
    IN_COLAB = True
except Exception:
    IN_COLAB = False

# -------------------------
# 1) Paths
# -------------------------
XLSX_PATH = "/content/drive/MyDrive/CommmedpatientTT.xlsx"
assert os.path.exists(XLSX_PATH), f"File not found: {XLSX_PATH}"

OUTDIR = "/content/figure5_outputs"
os.makedirs(OUTDIR, exist_ok=True)

# -------------------------
# 2) Load + column harmonization
# -------------------------
df0 = pd.read_excel(XLSX_PATH)
rename_map = {
    "HDL": "HDL-C",
    "LDL": "LDL-C",
    "HDL_C": "HDL-C",
    "LDL_C": "LDL-C",
    "BodyWeight": "BW",
    "Weight": "BW",
}
df0 = df0.rename(columns={k: v for k, v in rename_map.items() if k in df0.columns})

required_cols = ["age","height","BW","BMI","Hb","Hct","Alb","AST","ALT","ALP","Cre",
                 "Glu","TG","HDL-C","LDL-C","LH","FSH","CRP","TT"]
missing = [c for c in required_cols if c not in df0.columns]
assert len(missing) == 0, f"Missing columns: {missing}"

X_raw = df0[required_cols].copy()

# -------------------------
# 3) BMI reconstruction (fill only missing BMI)
# -------------------------
h = pd.to_numeric(X_raw["height"], errors="coerce")
w = pd.to_numeric(X_raw["BW"], errors="coerce")

height_is_cm = h.median(skipna=True) > 3.0
h_m = h / 100.0 if height_is_cm else h
bmi_from_hw = w / (h_m ** 2)

n_bmi_missing_before = int(X_raw["BMI"].isna().sum())
mask_bmi_na = X_raw["BMI"].isna()
X_raw.loc[mask_bmi_na, "BMI"] = bmi_from_hw.loc[mask_bmi_na]
n_bmi_missing_after = int(X_raw["BMI"].isna().sum())

print("\n[BMI reconstruction]")
print("height unit assumed:", "cm" if height_is_cm else "m")
print("BMI missing before:", n_bmi_missing_before)
print("BMI missing after :", n_bmi_missing_after)

# -------------------------
# 4) Subjects filter: complete age & BMI
# -------------------------
mask_subj = X_raw["age"].notna() & X_raw["BMI"].notna()
X_subj = X_raw.loc[mask_subj].copy()
N = len(X_subj)
print("\n[Subjects]")
print("N subjects (age & BMI complete):", N)  # 5854 expected

# -------------------------
# 5) Mean imputation + standardization for clustering
# -------------------------
imputer = SimpleImputer(strategy="mean")
scaler = StandardScaler()

X_imp = imputer.fit_transform(X_subj.values)
X_std = scaler.fit_transform(X_imp)

# -------------------------
# 6) KMeans (K=4 fixed)
# -------------------------
K = 4
RANDOM_STATE = 0
N_INIT = 50

kmeans = KMeans(n_clusters=K, random_state=RANDOM_STATE, n_init=N_INIT)
labels = kmeans.fit_predict(X_std)

sizes = pd.Series(labels).value_counts().sort_index()
print("\n[KMeans]")
print("K:", K, "random_state:", RANDOM_STATE, "n_init:", N_INIT)
print("Cluster sizes:", sizes.to_dict())

# attach cluster to subject-level dataframe (original scale after BMI fill)
df = X_subj.copy()
df["cluster"] = labels

# -------------------------
# 7) Utilities: LOWESS (optional) + piecewise breakpoint (no pwlf)
# -------------------------
def lowess_smooth(x, y, frac=0.25):
    """Return x_sorted, y_smooth. Uses statsmodels.lowess if available; else rolling median."""
    x = np.asarray(x, dtype=float)
    y = np.asarray(y, dtype=float)
    m = np.isfinite(x) & np.isfinite(y)
    x = x[m]; y = y[m]
    if len(x) < 50:
        xs = np.sort(x)
        return xs, np.interp(xs, x[np.argsort(x)], y[np.argsort(x)])
    order = np.argsort(x)
    x = x[order]; y = y[order]
    try:
        from statsmodels.nonparametric.smoothers_lowess import lowess
        sm = lowess(y, x, frac=frac, return_sorted=True)
        return sm[:,0], sm[:,1]
    except Exception:
        # fallback: rolling median on sorted x
        xs = x
        ys = y
        win = max(25, int(len(xs)*0.02))
        y_med = pd.Series(ys).rolling(win, center=True, min_periods=1).median().to_numpy()
        return xs, y_med

def piecewise_breakpoint(x, y, x_min=None, x_max=None, grid_step=0.5, min_n_each=200):
    """
    Simple two-segment piecewise linear fit by grid search over breakpoint.
    Returns dict with breakpoint, slopes, intercepts, SSE.
    """
    x = np.asarray(x, dtype=float)
    y = np.asarray(y, dtype=float)
    m = np.isfinite(x) & np.isfinite(y)
    x = x[m]; y = y[m]
    if len(x) < (min_n_each*2 + 10):
        return {"break": np.nan, "slope1": np.nan, "slope2": np.nan,
                "int1": np.nan, "int2": np.nan, "SSE": np.nan, "N": len(x)}

    if x_min is None: x_min = np.nanpercentile(x, 5)
    if x_max is None: x_max = np.nanpercentile(x, 95)

    # grid of candidate breakpoints
    bps = np.arange(x_min, x_max + 1e-9, grid_step)

    best = None
    for bp in bps:
        left = x < bp
        right = x >= bp
        if left.sum() < min_n_each or right.sum() < min_n_each:
            continue

        # fit y = a + b x by least squares
        Xl = np.column_stack([np.ones(left.sum()), x[left]])
        Xr = np.column_stack([np.ones(right.sum()), x[right]])
        bl = np.linalg.lstsq(Xl, y[left], rcond=None)[0]  # [a,b]
        br = np.linalg.lstsq(Xr, y[right], rcond=None)[0]

        yhat_l = Xl @ bl
        yhat_r = Xr @ br
        sse = np.sum((y[left] - yhat_l)**2) + np.sum((y[right] - yhat_r)**2)

        if (best is None) or (sse < best["SSE"]):
            best = {"break": float(bp),
                    "int1": float(bl[0]), "slope1": float(bl[1]),
                    "int2": float(br[0]), "slope2": float(br[1]),
                    "SSE": float(sse), "N": int(len(x)),
                    "N_left": int(left.sum()), "N_right": int(right.sum())}

    if best is None:
        return {"break": np.nan, "slope1": np.nan, "slope2": np.nan,
                "int1": np.nan, "int2": np.nan, "SSE": np.nan, "N": len(x)}
    return best

def spearman_xy(a, b):
    a = pd.to_numeric(a, errors="coerce")
    b = pd.to_numeric(b, errors="coerce")
    m = a.notna() & b.notna()
    if m.sum() < 3:
        return (np.nan, np.nan, int(m.sum()))
    r, p = spearmanr(a[m], b[m])
    return (float(r), float(p), int(m.sum()))

# -------------------------
# 8) Compute statistics for Figure 5 (overall + clusters 0&3)
# -------------------------
# Overall: TT~age, CRP~age, Cre~age
overall_stats = []
for var in ["TT","CRP","Cre"]:
    r, p, n = spearman_xy(df["age"], df[var])
    overall_stats.append({"subset":"overall", "x":"age", "y":var, "N":n, "spearman_r":r, "p_value":p})
overall_stats = pd.DataFrame(overall_stats)

# Overall TT–CRP correlation
r_tc, p_tc, n_tc = spearman_xy(df["TT"], df["CRP"])
ttcrp_overall = pd.DataFrame([{"subset":"overall", "x":"TT", "y":"CRP", "N":n_tc, "spearman_r":r_tc, "p_value":p_tc}])

# Clusters 0&3 subset
df03 = df[df["cluster"].isin([0,3])].copy()

# piecewise in clusters 0&3 vs age for each y
pw_records = []
for y in ["TT","CRP","Cre"]:
    res = piecewise_breakpoint(df03["age"], df03[y], grid_step=0.5, min_n_each=200)
    res.update({"subset":"clusters0_3", "x":"age", "y":y})
    pw_records.append(res)
piecewise03 = pd.DataFrame(pw_records)

# Spearman in clusters 0&3: TT–CRP, TT–Cre
r_tcrp_03, p_tcrp_03, n_tcrp_03 = spearman_xy(df03["TT"], df03["CRP"])
r_tcre_03, p_tcre_03, n_tcre_03 = spearman_xy(df03["TT"], df03["Cre"])
spearman03 = pd.DataFrame([
    {"subset":"clusters0_3", "x":"TT", "y":"CRP", "N":n_tcrp_03, "spearman_r":r_tcrp_03, "p_value":p_tcrp_03},
    {"subset":"clusters0_3", "x":"TT", "y":"Cre", "N":n_tcre_03, "spearman_r":r_tcre_03, "p_value":p_tcre_03},
])

# save stats
overall_stats_path = os.path.join(OUTDIR, "figure5_spearman_overall_age.csv")
ttcrp_overall_path = os.path.join(OUTDIR, "figure5_spearman_overall_TT_CRP.csv")
piecewise03_path = os.path.join(OUTDIR, "figure5_piecewise_clusters0_3.csv")
spearman03_path = os.path.join(OUTDIR, "figure5_spearman_clusters0_3.csv")

overall_stats.to_csv(overall_stats_path, index=False)
ttcrp_overall.to_csv(ttcrp_overall_path, index=False)
piecewise03.to_csv(piecewise03_path, index=False)
spearman03.to_csv(spearman03_path, index=False)

print("\nSaved stats:")
print(overall_stats_path)
print(ttcrp_overall_path)
print(piecewise03_path)
print(spearman03_path)

# -------------------------
# 9) Plot Figure 5 (a–f): overall (a–c), clusters 0&3 (d–f)
# -------------------------
def scatter_with_lowess(ax, dfx, xcol, ycol, title, ylab, xlab="Age (years)", marker="o", alpha=0.15, s=8):
    x = pd.to_numeric(dfx[xcol], errors="coerce")
    y = pd.to_numeric(dfx[ycol], errors="coerce")
    m = x.notna() & y.notna()
    ax.scatter(x[m], y[m], s=s, alpha=alpha, marker=marker)
    xs, ys = lowess_smooth(x[m], y[m], frac=0.25)
    ax.plot(xs, ys, linewidth=2)
    ax.set_title(title)
    ax.set_xlabel(xlab)
    ax.set_ylabel(ylab)

# Set up figure
fig, axes = plt.subplots(2, 3, figsize=(14, 8), constrained_layout=True)

# (a–c) overall
scatter_with_lowess(axes[0,0], df, "age", "TT", "(a) Overall", "TT (ng/mL)")
scatter_with_lowess(axes[0,1], df, "age", "CRP", "(b) Overall", "CRP (mg/dL)")
scatter_with_lowess(axes[0,2], df, "age", "Cre", "(c) Overall", "Creatinine (mg/dL)")

# (d–f) clusters 0&3, markers by cluster
df_c0 = df[df["cluster"] == 0].copy()
df_c3 = df[df["cluster"] == 3].copy()

# (d) TT
scatter_with_lowess(axes[1,0], df_c0, "age", "TT", "(d) Clusters 0 & 3", "TT (ng/mL)", marker="o", alpha=0.20, s=9)
scatter_with_lowess(axes[1,0], df_c3, "age", "TT", "", "", marker="x", alpha=0.20, s=12)
# piecewise line overlay (clusters0_3 combined)
pw_tt = piecewise03[piecewise03["y"]=="TT"].iloc[0].to_dict()
if np.isfinite(pw_tt.get("break", np.nan)):
    bp = pw_tt["break"]
    # left segment
    x_min = np.nanmin(df03["age"]); x_max = np.nanmax(df03["age"])
    xl = np.linspace(x_min, bp, 200)
    xr = np.linspace(bp, x_max, 200)
    yl = pw_tt["int1"] + pw_tt["slope1"] * xl
    yr = pw_tt["int2"] + pw_tt["slope2"] * xr
    axes[1,0].plot(xl, yl, linewidth=2)
    axes[1,0].plot(xr, yr, linewidth=2)
    axes[1,0].axvline(bp, linestyle="--", linewidth=1)

# (e) CRP
scatter_with_lowess(axes[1,1], df_c0, "age", "CRP", "(e) Clusters 0 & 3", "CRP (mg/dL)", marker="o", alpha=0.20, s=9)
scatter_with_lowess(axes[1,1], df_c3, "age", "CRP", "", "", marker="x", alpha=0.20, s=12)
pw_crp = piecewise03[piecewise03["y"]=="CRP"].iloc[0].to_dict()
if np.isfinite(pw_crp.get("break", np.nan)):
    bp = pw_crp["break"]
    x_min = np.nanmin(df03["age"]); x_max = np.nanmax(df03["age"])
    xl = np.linspace(x_min, bp, 200)
    xr = np.linspace(bp, x_max, 200)
    yl = pw_crp["int1"] + pw_crp["slope1"] * xl
    yr = pw_crp["int2"] + pw_crp["slope2"] * xr
    axes[1,1].plot(xl, yl, linewidth=2)
    axes[1,1].plot(xr, yr, linewidth=2)
    axes[1,1].axvline(bp, linestyle="--", linewidth=1)

# (f) Cre
scatter_with_lowess(axes[1,2], df_c0, "age", "Cre", "(f) Clusters 0 & 3", "Creatinine (mg/dL)", marker="o", alpha=0.20, s=9)
scatter_with_lowess(axes[1,2], df_c3, "age", "Cre", "", "", marker="x", alpha=0.20, s=12)
pw_cre = piecewise03[piecewise03["y"]=="Cre"].iloc[0].to_dict()
if np.isfinite(pw_cre.get("break", np.nan)):
    bp = pw_cre["break"]
    x_min = np.nanmin(df03["age"]); x_max = np.nanmax(df03["age"])
    xl = np.linspace(x_min, bp, 200)
    xr = np.linspace(bp, x_max, 200)
    yl = pw_cre["int1"] + pw_cre["slope1"] * xl
    yr = pw_cre["int2"] + pw_cre["slope2"] * xr
    axes[1,2].plot(xl, yl, linewidth=2)
    axes[1,2].plot(xr, yr, linewidth=2)
    axes[1,2].axvline(bp, linestyle="--", linewidth=1)

# Add a minimal legend-like note (without cluttering)
fig.suptitle("Figure 5. Age-related patterns of TT, CRP, and creatinine (overall and clusters 0 & 3)", y=1.02)

fig_path_png = os.path.join(OUTDIR, "Figure5.png")
fig_path_pdf = os.path.join(OUTDIR, "Figure5.pdf")
plt.savefig(fig_path_png, dpi=300, bbox_inches="tight")
plt.savefig(fig_path_pdf, bbox_inches="tight")
plt.show()

print("\nSaved figures:")
print(fig_path_png)
print(fig_path_pdf)

# -------------------------
# 10) Download (no ZIP)
# -------------------------
if IN_COLAB:
    for p in [fig_path_png, fig_path_pdf, overall_stats_path, ttcrp_overall_path, piecewise03_path, spearman03_path]:
        files.download(p)
else:
    print("\nNot in Colab. Files are in:", OUTDIR)
